src/FixedDataTableCellGroup.js

```diff
@@ -76,25 +87,70 @@ var FixedDataTableCellGroupImpl = React.createClass({
     }, false);
 
     var currentPosition = 0;
-    for (var i = 0, j = columns.length; i < j; i++) {
+    var count = 0;
+
+    var newColumnsToRender = new Array(props.maxVisibleColumns);
+    var positions = new Array(props.maxVisibleColumns);
+
+    for (var i = 0; i < columns.length; i++) {
       var columnProps = columns[i].props;
       var recycable = columnProps.allowCellsRecycling && !isColumnReordering;
       if (!recycable || (
-            currentPosition - props.left <= props.width &&
-            currentPosition - props.left + columnProps.width >= 0)) {
-        var key = columnProps.columnKey || 'cell_' + i;
-        cells[i] = this._renderCell(
-          props.rowIndex,
-          props.rowHeight,
-          columnProps,
-          currentPosition,
-          key,
-          contentWidth,
-          isColumnReordering
-        );
+          currentPosition - props.left <= props.width &&
+          currentPosition - props.left + columnProps.width >= 0)) {
+        positions[i] = currentPosition;
+        newColumnsToRender[count++] = i;
       }
+
       currentPosition += columnProps.width;
     }
+
+    //TODO move this recycle logic into main state
+    const newColumnsSet = new Set(newColumnsToRender);
+    const oldColumnsSet = new Set(this._columnsToRender);
+    const indexes = [];
+
+    for (var i = props.maxVisibleColumns; i >= 0; i--) {
+      var column = this._columnsToRender[i];
+      if (!column || !newColumnsSet.has(column)) {
```

Comment:
Above it looks like you put an index in newColumnsToRender, but here your refer to it as a column...  I'm pretty sure it's an index, but it's a bit confusing to read.